### PR TITLE
Licence extractor

### LINKED ISSUE #81: Define dataclasses for extractor results.
Currently the individual extractors return dynamic a dictionaries which somewhat follow the expected api output structure and are combined and enriched in the `MetadataManager` class. Hence, the API return models currently leak through the manager class into the extractors.
We should define a dataclass for the extractor return values which is independent of the output model. The expected output model can then be build either in the manager class, (at a later stage) even in the endpoint implementation.

AC:
- Extractors have a uniform extraction return type and output model is contained in manager or api endpoint implementation.

Note:
- If an extractor needs to deviate from a uniform output structure (e.g. because it provides extra information about how it came to its X star rating), then it should declare its own return structure declaratively. It would then also be the managers responsibility how to deal with this extra information.

### PR DESCRIPTION
**Pull request by MRuecklCC**
- Provide a simple licence detection method by scanning the html content for strings like `CC BY`, `CC BY-SA`, `Creative Commons Attributions` etc.
- Defines a new abstract base extractor interface. This allows keeping the current extractor hierarchy via `MetadataBase` which can gradually be migrated to the new base class in the future.
- The `/extract` endpoint now takes an additional query parameter `/extract?extra=(true|false)` (defaults to false). If set to true, extra information of the extractors is provided in the response body. The structure of this extra information is extractor specific and defined via the generic parameter `T` in the extractor classes.

Resolves #81 
Works towards #73 
Works towards #83

### CODE REVIEW

**Comment by RMeissnerCC** [comment]
Careful with the urls.txt: this is basically IP of WLO - I thus vote against merging it with the current state of the urls.txt. Please place it elsewhere, where it is not directly public.

**Comment by RMeissnerCC** [comment]
I have a hard time testing this, because for some reason nothing wants to build locally :/ I presume we need another github action creating :dev images next to the :latest tags.

File: tests/extractors/licence_test.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    pattern = "(?P<foo>(CC BY))|(CC BY-SA)|(Creative Commons Zero|CC0)"
+    pattern = re.compile(pattern)
+    print(pattern.flags, pattern.groups, pattern.groupindex)
+    for i in pattern.findall(html):
+        print(i)
+
+    # print(pattern.findall(html))
+
+
+# fixme: remove or activate. This reads through a set of splash responses and analyzes them to get an idea how many
```
Comment (RMeissnerCC) [inline tests/extractors/licence_test.py:50]: Instead of commenting, either skip it with pytest - so it appears in pytest and refactoring - or delete it.

**Review by RMeissnerCC — CHANGES_REQUESTED** [review]
Some nice changes. I have issues getting it run, perhaps we can look at it together on monday.

File: tests/integration/api_integration_test.py [hunk 2]
```diff
@@ -22,6 +23,9 @@ async def client() -> AsyncClient:
     cache_db = Path(__file__).parent.parent / "meta-lookup-cache.db"
     if os.path.exists(cache_db):
         os.remove(cache_db)
+    from databases import Database
+
+    cache_backend.database = Database("sqlite://./meta-lookup-cache.db")
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply tests/integration/api_integration_test.py:28]: SQLite will automatically create it if it does not exists.

File: tests/extractors/licence_test.py (hunk 1 above)
Comment (MRuecklCC) [reply tests/extractors/licence_test.py:50]: Yes, maybe we can have a quick chat about that today. It also somewhat touches the urls.txt topic.

File: src/core/metadata_manager.py [hunk 3]
```diff
@@ -98,14 +111,17 @@ async def extract(self, message: Input) -> Output:
...
             try:
                 self.logger.debug(f"Extracting {extractor.__class__.__name__}.")
                 with runtime() as t:
-                    _, _, stars, explanation = await extractor.start(site=site)
+                    stars, explanation, extra_data = await extractor.extract(site=site, executor=self.process_pool)
                 self.logger.info(f"Extracted {extractor.__class__.__name__} in {t():5.2f}s.")
-                return MetadataTags(stars=stars, explanation=explanation)
+                if extra:
+                    return MetadataTags(stars=stars, explanation=explanation, extra=extra_data)
+                else:
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/core/metadata_manager.py]: fixed

File: src/features/accessibility.py [hunk 4]
```diff
@@ -4,17 +4,21 @@
...
+from core.extractor import NO_EXPLANATION, MetadataBase
 from core.website_manager import WebsiteData
 from lib.constants import ACCESSIBILITY
 from lib.settings import LIGHTHOUSE_TIMEOUT, LIGHTHOUSE_URL
 
 _DESKTOP = "desktop"
 _MOBILE = "mobile"
 
+_LIGHTHOUSE_SCORE_SUITABLE = "Lighthouse accessibility score is suitable"
+_LIGHTHOUSE_SCORE_TO_LOW = "Lighthouse accessibility score is to low"
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/features/accessibility.py]: fixed

[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/features/accessibility.py]: fixed

File: tests/conftest.py [hunk 5]
```diff
@@ -11,5 +19,41 @@ def pytest_configure(config):
...
+
+
+@pytest.fixture
+def splash_samples() -> list[SplashResponse]:
+    """Load splash responses from test resources."""
+
+    def samples():
+        for _, _, files in os.walk(Path(__file__).parent / "resources"):
+            for file in files:
+                if file.startswith("splash-sample"):
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply tests/conftest.py]: removed that code and openend #107

[the remark this answers is not in the document]
Comment (MRuecklCC) [reply tests/conftest.py]: removed that code and opened #107

File: tests/integration/api_integration_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply tests/integration/api_integration_test.py:28]: SQLite creates it if it does not exist

[the remark this answers is not in the document]
Comment (MRuecklCC) [reply tests/integration/api_integration_test.py:28]: SQLite creates it if it does not exist

File: tests/extractors/licence_test.py (hunk 1 above)
Comment (MRuecklCC) [reply tests/extractors/licence_test.py]: removed it

File: src/features/licence.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    key = "licence"
+
+    def __init__(self):
+        self.ranking: dict[Licence, StarCase] = {
+            # least restrictive?
+            Licence.CC0: StarCase.FIVE,
+            Licence.CC_BY: StarCase.FIVE,
+            Licence.CC_BY_SA: StarCase.FIVE,
+            # pretty standard and good?
+            # fixme: eventually we want to rang CC ND licences lower,
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/features/licence.py:56]: I added this as a todo to the Jira ticket. Would still like to go ahead and merge this. If they want the ranking to be changed or other licences added I would do that in an extra PR!

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/features/licence.py:56]: Good point. Potentially the crawler has some solution here, already.

**Review by RMeissnerCC — APPROVED** [review]
Great :) Plz merge

**Pull request merged**